feat: add names to args at invocation and allow omitting optional args

westonpace opened the pull request:
BREAKING CHANGE: optional arguments may now be omitted.  However, if they are specified then they MUST be named.

westonpace commented:
Hopefully this breaking change won't be quite so contentious.  I imagine there are other ways we could implement this as well (e.g. we could require all arguments to be named).  Or, if we must, we could open the door to optional argument inference strategies (though I'd rather not go down that road).  I'm open to alternative suggestions.

I'm filing this because it is becoming a problem in Acero.  Acero currently expects all arguments (including optional arguments) to always be specified (I believe this is the validator's interpretation as well).  Unfortunately, neither the Isthmus or DuckDb producers behave in this way.

As an example:

```
isthmus -c "CREATE TABLE T1(foo int)" "SELECT foo + 1 FROM T1"
{
  ...
          "expressions": [{
            "scalarFunction": {
              ...
              "arguments": [{"value": { ... }}, {"value": { ... }}], // Missing overflow
  ...

}
```

westonpace commented:
As an added bonus, and something that was brought up in the last community sync, this change should enable new optional arguments to be added to a function without breaking backwards compatibility.

jacques-n commented:
I'm strongly inclined to pull the options out of the arguments array. It's a bigger breaking change but seems cleaner. When @jvanstraten mentioned names on the sync, that was more what I was expecting.

westonpace commented:
> I'm strongly inclined to pull the options out of the arguments array.

That's fine with me.  I don't think it's any more work either way for us.  Proto3 actually supports [maps](https://developers.google.com/protocol-buffers/docs/proto3#maps).  Would we want to express the named options as a map then?

jvanstraten commented:
> I'm strongly inclined to pull the options out of the arguments array. It's a bigger breaking change but seems cleaner. When @jvanstraten mentioned names on the sync, that was more what I was expecting.

+1, that's what I was intending, but I just hadn't gotten around to proposing something yet between all the reviewing and validator work. Likewise for lambdas by the way.

Longer-term, if we end up embracing the meta-type system that formed as part of my type derivation work, required enumeration arguments and type arguments could both be naturally expressed with a metavalue. You'd also get static boolean, integer, and string arguments for free, all usable within the type derivation system. They'd basically be the generalization of what most languages call template or generic arguments.

I'd then be inclined to just pull them into a different `repeated` as well, so you'd end up with something like

```proto
// A scalar function call.
message ScalarFunction {
  // Points to a function_anchor defined in this plan, which must refer
  // to a scalar function in the associated YAML file. Required; avoid
  // using anchor/reference zero.
  uint32 function_reference = 1;
  
  // Options to specify behavior for corner cases, or leave behavior unspecified
  // if not necessary.
  repeated FunctionOption options = 2;
  message FunctionOption {
    // Name of the option to set. If the consumer does not recognize the
    // option, it must reject the plan. The name is matched case-insensitively.
    string name = 1;
    
    // List of behavior options allowed by the producer. At least one must be
    // specified; to leave an option unspecified, simply don't add an entry to
    // `options`. The consumer must use the first option from the list that it
    // supports. If the consumer supports none of the specified options, it
    // must reject the plan. The name is matched case-insensitively.
    repeated string preference = 2;
  }
  
  // Generic parameters to configure the function statically. Must match the
  // signature of the function exactly.
  repeated Metavalue parameters = 3;
  
  // Runtime arguments passed to the function. The data types of the values
  // returned by the expressions must match the signature of the function
  // exactly. Constant arguments must be mapped to a Literal expression.
  repeated Expression arguments = 4;
  
  // Must be (redundantly) set to the return type of the function, exactly as
  // derived using the declaration in the extension.
  Type output_type = 5;
}
```

where `Metavalue` looks something like [this](https://github.com/substrait-io/substrait-validator/blob/a3a2fbc07761a61ec0756aec8470be005bee8e12/proto/substrait/validator/type_system.proto#L180-L203), but without the validator-specific `unresolved` option.

(note on case insensitivity: I dislike it personally, but dislike inconsistency more)

> Proto3 actually supports maps. Would we want to express the named options as a map then?

I'd prefer not to because A) I'm not sure how `prost` and other third-party protobuf implementations would deal with it (if at all) and B) it's essentially just sugar for

```proto
message MapFieldEntry {
  key_type key = 1;
  value_type value = 2;
}

repeated MapFieldEntry map_field = N;
```

anyway (as copypasted from the protobuf docs).

---

An important thing we need to change here still is that optional enumeration arguments need to be removed from the compound name signatures of functions, otherwise adding such arguments would *still* be a breaking change.

westonpace commented:
I've switched to the message format proposed by @jvanstraten and updated the spec around function signatures to specifically mention that options are not part of the signature.

jvanstraten reviewed (dismissed):
The implementation and semantics look good to me. I'm neutral about the documentation; I think it covers all the semantics but I think it might be easy to miss how different the various argument types are and behave, especially because the "argument type" concept doesn't translate well to other languages (at least none that I know). This is not necessarily a critique of this PR but more of the function docs in general though, and that can be improved later.

jacques-n commented on proto/substrait/algebra.proto:
```diff
@@ -648,11 +647,27 @@ message Expression {
+    // unspecified if the consumer does not need specific behavior in these
+    // cases.
+    repeated FunctionOption options = 5;
+    message FunctionOption {
```
Why declare this in each function type? FunctionArgument is external and it seems like this should similarly be so.

westonpace replied:
Good point.  I've moved this so it is declared once, next to `FunctionArgument`.

jacques-n reviewed (changes requested):
My expectation here is this should also include a change to the yaml schema where optional arguments are not specified in the same way as other argument types. I don't see that identified here nor do I see any change to the yaml schema.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/substrait-io/substrait?pullRequest=342) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/substrait-io/substrait?pullRequest=342) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/substrait-io/substrait?pullRequest=342) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/substrait-io/substrait?pullRequest=342) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/substrait-io/substrait?pullRequest=342) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/substrait-io/substrait?pullRequest=342) it.</sub>

westonpace commented:
> My expectation here is this should also include a change to the yaml schema where optional arguments are not specified in the same way as other argument types. I don't see that identified here nor do I see any change to the yaml schema.

@jacques-n That was a significantly more extensive change.  PTAL and let me know if this is what you were thinking.

bkietz commented on site/docs/extensions/generate_function_docs.py:
```diff
@@ -37,6 +37,8 @@ def write_markdown(file_obj: dict, file_name: str) -> None:
             options_list = []
 
             for count, impls in enumerate(implementations_list):
+                if "args" not in impls:
```
AFAICT, this is not reading the "options" field of `impls` so we won't get any doc from that. Furthermore, I would expect `options_list` to be constructed from each `impls["options"]` entry since each impl may define its own options. Instead, since that variable is outside the loop it is accumulating the union of all options from anything in `implementation_list`- we haven't noticed because most impls of a single function have identical options.

It might be desirous to *require* that all impls have identical options, possibly by hoisting options to a property of the function rather than of the impl. That adds some validation overhead when multiple function declarations share a name- what should be the behavior if regular "add" and decimal "add" declare different options? I'd say keeping options as a property of `impls[n]` is probably preferable

westonpace replied:
I'll fix this up so we generate docs for options.  Yes, we will want options to be per-impl I think.  There is some precedent for this already.  Integer arithmetic usually has overflow for example while fp arithmetic had rounding mode.

westonpace replied:
Ok.  I restored it to what it used to be.  However, I left it per-impl.  This can go wrong as it does here:

![image](https://user-images.githubusercontent.com/1696093/196623089-21e38045-6b31-4fbd-8050-e47707db5a41.png)

However, that can be left for a future issue.

richtia commented:
@pdet

jacques-n commented on extensions/functions_aggregate_generic.yaml:
```diff
@@ -5,22 +5,21 @@ aggregate_functions:
     description: "Count a set of records (not field referenced)"
     impls:
-      - args:
+      - options:
```
did you consider making these maps instead of arrays? That way the name is more formally required purely by structure.

westonpace replied:
That's a good idea and solves my other naming problem.  I'll give that a try.

westonpace replied:
Done.

jacques-n commented on extensions/functions_datetime.yaml:
```diff
@@ -6,30 +6,26 @@ scalar_functions:
         return: i64
       - args:
-          - options: [ YEAR, MONTH, DAY, SECOND ]
+          - values: [ YEAR, MONTH, DAY, SECOND ]
```
I'm surprised by your change here. Are you renaming required enumeration argument types?

westonpace replied:
My choices were to leave this alone and have it be out of sync with options (`values` vs `options`) or to have the word "options" doubly used in options `options::options`.  However, if `options` is a map then this is a moot point so I will revert this change.

westonpace replied:
Actually I take that back.  The map change doesn't fix this.  So basically we can have approach 1:

```
args:
 - name: some_required_enum
   options: [ X, Y ]
options:
 some_optional_enum:
  options: [ A, B ]
```

Which is consistent but has some shadowing with the double options.  Or we can have approach 2:

```
args:
 - name: some_required_enum
   options: [ X, Y ]
options:
 some_optional_enum:
  values: [ A, B ]
```

This avoids the shadowing but is inconsistent (options vs. values).  Or we can have approach 3:

```
args:
 - name: some_required_enum
   values: [ X, Y ]
options:
 some_optional_enum:
   values: [ A, B ]
```

This avoids the shadowing and is consistent but is more of a breaking change.

I'm happy to go with any of the above choices.

westonpace replied:
I have currently implemented approach 2.

jacques-n commented on site/docs/expressions/scalar_functions.md:
```diff
@@ -8,7 +8,7 @@ A function is a scalar function if that function takes in values from a single r
 | Deterministic          | Whether this function is expected to reproduce the same output when it is invoked multiple times with the same input. This informs a plan consumer on whether it can constant-reduce the defined function. An example would be a random() function, which is typically expected to be evaluated repeatedly despite having the same set of inputs. | Optional, defaults to true.      ...
 | Session Dependent      | Whether this function is influenced by the session context it is invoked within. For example, a function may be influenced by a user who is invoking the function, the time zone of a session, or some other non-obvious parameter. This can inform caching systems on whether a particular function is cacheable. | Optional, defaults to false.        |
-| Variadic Behavior      | Whether the last argument of the function is variadic or a single argument.  If variadic, the argument can optionally have a lower bound (minimum number of instances) and an upper bound (maximum number of instances). | Optional, defaults to single value. |
+| Variadic Behavior      | Whether the last required argument of the function is variadic or a single argument.  If variadic, the argument can optionally have a lower bound (minimum number of instances) and an upper bound (maximum number of instances). | Optional, defaults to single value. |
```
this change seems unnecessary.

westonpace replied:
Reverted.

jacques-n commented on site/docs/expressions/scalar_functions.md:
```diff
@@ -19,11 +19,14 @@ A function is a scalar function if that function takes in values from a single r
 ## Argument Types
 
 There are four main types of arguments: value arguments, type arguments, required enumerations, and optional enumerations.
+The first three types of arguments are considered positional arguments and must be specified in every invocation of the
```
Doesn't feel quite right with new pattern.

westonpace replied:
I reworded this entire section.  Please review.

jacques-n reviewed (commented):
Good direction. Couple of couples...

bkietz commented on site/docs/expressions/scalar_functions.md:
```diff
@@ -47,12 +46,20 @@ There are four main types of arguments: value arguments, type arguments, require
+
+A producer may specify multiple vlaues for an option.  If the producer does so then the consumer must deliver the first behavior in the list of values that the consumer is capable of delivering.  For example, considering overflow as defined above, if a producer specified `[ERROR, SATURATE]` then the consumer must deliver `ERROR` if it is capable of doing so.  If it is not then it may deliver `SAT ...
+
+#### Optional Properties
```
There's some confusion here since it's not completely clear whether we're talking about the properties of an `algebra.proto::substrait.FunctionOption` message or about the properties of a `simple_extensions_schema.yaml::options` declaration (in general, I think this document would be improved by explicitly listing the substrait message to which these `Property, Description, Required` tables apply).

I would delete `### Optional Precedence``` and hoist this section above the paragraph explaining precedence. That paragraph can then serve as the explanation as to why an `algebra.proto::substrait.FunctionOption` has a list of values instead of a single value.

westonpace replied:
> There's some confusion here since it's not completely clear whether we're talking about the properties of an algebra.proto::substrait.FunctionOption message or about the properties of a simple_extensions_schema.yaml::options declaration (in general, I think this document would be improved by explicitly listing the substrait message to which these Property, Description, Required tables apply).

I had the same confusion.  I think this entire section should probably be moved somewhere else.  This page is describing what belongs in a plan and this section is describing what belongs in extension files.  Those two blocks of documentation should be on separate pages.  The page describing what belongs in a plan can talk about precedence/preference.  The page describing what belongs in the extension yamls can describe the "set of possible values".  I would rather keep this as is and fix this in a future PR.  The proposed change isn't quite right either I think.

bkietz commented:
Other than these nits, LGTM

westonpace commented on proto/substrait/algebra.proto:
```diff
@@ -490,11 +489,28 @@ message FunctionArgument {
     }
+    reserved 2;
   }
 }
```
@jacques-n Any opinion on this suggestion?  This would get rid of `message Enum` entirely which seems a lot simpler.

westonpace replied:
@cpcloud @jacques-n friendly ping.  I feel this PR has stalled a bit and I would like to get it through.  However, we should evaluate this carefully as I don't think we can easily revisit in the future.  Any second opinions on this suggestion?

westonpace replied:
@cpcloud @jacques-n ping again.

cpcloud replied:
This is a pretty big change, but I am in favor of it.

westonpace replied:
I have made this change.

westonpace commented:
I was going to interpret [this comment](https://github.com/substrait-io/substrait/pull/342#discussion_r1015637716) as a second SMC approval and merge this :)

However, it seems I cannot merge while a review is left in the "change requested" state.  @jacques-n do you know if you will be able to look at this anytime soon?

westonpace commented:
> any chance someone on your team can pick up the downstream changes for validator and java?

I'll see what I can do.
